File: Euler_274_Divisibility_Multipliers/java/euler274DivisibilityMultipliers.java
public class euler274DivisibilityMultipliers {
    public static void main(String[] args) {
        int limit = Integer.parseInt(args[0]);
        long sumOfDivisibilityMultipliers = 0;

        for (int i = 1; i < limit; i++) {
            if (areCoprime(i, 10) && isPrime(i)) {
                int x = findDivisibilityMultiplier(i, limit);
                sumOfDivisibilityMultipliers += x;
            }
        }

        System.out.println("Answer: " + sumOfDivisibilityMultipliers);
    }

    /** Check if two numbers are coprimes. */
    private static boolean areCoprime(int a, int b) {
        return gcd(a, b) == 1;
    }

    /** Calculate the greatest common divisor (GCD) of two numbers. */
    private static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    /** Check if a number is prime. */
    private static boolean isPrime(int n) {
        if (n <= 1) {
            return false;
        }
        for (int i = 2; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                return false;
            }
        }
        return true;
    }

    /** Find the divisibility multiplier for a prime number. */
    private static int findDivisibilityMultiplier(int prime, int limit) {
        for (int x = 1; x < prime; x++) {
            boolean found = true;

            for (int n = 1; n < limit; n++) {
                int f_n = getFunctionValue(n, x);

                if ((f_n % prime != 0 && n % prime == 0) ||
                        (f_n % prime == 0 && n % prime != 0)) {
                    found = false;
                    break;
                }
            }

            if (found) {
                return x;
            }
        }

        // Default value if no x is found (should not happen for primes).
        return -1;
    }

    /** Calculate the value of the given function f(n). */
    private static int getFunctionValue(int n, int x) {
        int lastDigit = n % 10;
        int remainingDigits = n / 10;
        return remainingDigits + (lastDigit * x);
    }
}

File: Euler_274_Divisibility_Multipliers/go/euler-274-divisibility-multipliers.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

// Function to check if two numbers are coprime.
func areCoprime(a, b int) bool {
	return gcd(a, b) == 1
}

// Function to calculate the greatest common divisor (GCD) of two numbers.
func gcd(a, b int) int {
	for b != 0 {
		temp := b
		b = a % b
		a = temp
	}
	return a
}

// Function to check if a number is prime.
func isPrime(n int) bool {
	if n <= 1 {
		return false
	}
	for i := 2; i <= int(math.Sqrt(float64(n))); i++ {
		if n%i == 0 {
			return false
		}
	}
	return true
}

// Function to find the divisibility multiplier for a prime number.
func findDivisibilityMultiplier(prime, limit int) int {
	for x := 1; x < prime; x++ {
		found := true

		for n := 1; n < limit; n++ {
			fN := getFunctionValue(n, x)

			if (fN%prime != 0 && n%prime == 0) || (fN%prime == 0 && n%prime != 0) {
				found = false
				break
			}
		}

		if found {
			return x
		}
	}

	// Default value if no x is found (should not happen for primes).
	return -1
}

// Function to calculate the value of the given function f(n).
func getFunctionValue(n, x int) int {
	lastDigit := n % 10
	remainingDigits := n / 10
	return remainingDigits + (lastDigit * x)
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run euler-274-divisibility-multipliers.go <limit>")
		return
	}

	limit, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println("Invalid input. Please provide a valid integer limit.")
		return
	}

	sumOfDivisibilityMultipliers := 0

	for i := 1; i < limit; i++ {
		if areCoprime(i, 10) && isPrime(i) {
			x := findDivisibilityMultiplier(i, limit)
			sumOfDivisibilityMultipliers += x
		}
	}

	fmt.Println("Answer:", sumOfDivisibilityMultipliers)
}
